mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -2024,25 +2058,50 @@ int main(int argc, char **argv) {
     mod = miopenModule;
   }
 
+  assert(mod == module);
+
+
+  // Compute set of call-graph root nodes;  they're the ones we need to
+  // call from main().  Start with all nodes, then erase the ones that
+  // have edges to them.
+  CallGraph cg(mod);
+  llvm::DenseSet<CallGraphNode*> roots(cg.begin(), cg.end());
+  for (auto &node : roots)
+    for (auto &edge : *node)
+      roots.erase(edge.getTarget());
+
+  // Make KernelIFs for the roots, to pass to populateHostHarnessLogic().
+  SmallVector<KernelIF, 8> rootIFs;
+  for (auto node : roots) {
+    FuncOp func = dyn_cast<FuncOp>(node->getCallableRegion()->getParentOp());
+    rootIFs.emplace_back(func);
+  }
+
+  // Sort by name to be deterministic (because DenseSet above).
+  llvm::sort(rootIFs, [&](KernelIF &a, KernelIF &b) {
+                        return a.func.getName() < b.func.getName();
+                      });
+
   if (testFuncNameVal.empty()) {
     mod.walk([&](FuncOp func) -> WalkResult {
       if (func->hasAttr("kernel")) {
         kernels.emplace_back(func);
       }
       return WalkResult::advance();
     });
-  } else {
```

Comment:
Don't treat -fut as a kernel.  If it's naturally one, it'll be one.  If it's naturally a root, it'll be one.  I'm not sure which cases -fut gives us now.

Reply:
There could be many kernels and many roots in the file, but suppose we only want to test 1 of them.

Reply:
Yeah, I don't have a good answer for that yet.  I think we want to make -fut be a root, and the only root, but not a kernel unless it has the "kernel" attribute or there are no other kernels.

Reply:
I think it also makes sense for -fut to be mutually exclusive with kernel generation.